src/runner.rs

```diff
@@ -6,511 +6,183 @@
 //!
 //! The main methods are `Runner::run` and `Runner::result`.
 
+use std::mem;
+use std::panic;
 
-use context::*;
-use events;
-use events::Event;
-use example_result;
-use example_result::ExampleResult;
+use example_report::ExampleReport;
+use context_report::ContextReport;
+use visitor::Visitor;
+use events::{Event, EventHandler};
 
-pub type RunnerResult = Result<TestReport, TestReport>;
+use suite::Suite;
+use context::Context;
+use example::Example;
+use context::ContextMember;
 
 /// Handlers is a separate struct which only holds the registered handlers.
 /// This is useful to Runner so that its recursive call doesn't have to keep a refernce to `self`
 #[derive(Default)]
 struct Handlers<'a> {
-    handlers: Vec<&'a mut events::EventHandler>,
+    handlers: Vec<&'a mut EventHandler>,
 }
 
 impl<'a> Handlers<'a> {
-    fn broadcast(&mut self, event: &events::Event) {
+    fn broadcast(&mut self, event: &Event) {
         for h in &mut self.handlers {
             h.trigger(event)
         }
     }
 }
 
-pub struct Runner<'a> {
-    describe: Context<'a>,
-    report: Option<RunnerResult>,
+pub struct Runner<'a, T>
+    where T: 'a
+{
+    suite: Option<Suite<'a, T>>,
+    environments: Vec<T>,
+    report: ContextReport,
     handlers: Handlers<'a>,
 }
 
-impl<'a> Runner<'a> {
-    pub fn new(context: Context<'a>) -> Runner<'a> {
+impl<'a, T> Runner<'a, T> {
+    pub fn new(suite: Suite<'a, T>, environment: T) -> Runner<'a, T> {
         Runner {
-            describe: context,
+            suite: Some(suite),
+            environments: vec![environment],
+            report: ContextReport::default(),
             handlers: Handlers::default(),
-            report: None,
         }
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Clone, Copy, Default)]
-pub struct TestReport {
-    pub total_tests: u32,
-    pub success_count: u32,
-    pub error_count: u32,
-}
-
-impl<'a> Runner<'a> {
-
-    #[cfg_attr(feature="clippy", allow(redundant_closure))]
-    fn run_test(mut test_function: Box<TestFunction>)
-                -> ExampleResult {
-
-        use std::panic::{catch_unwind, AssertUnwindSafe};
-        use example_result;
-
-        let res = catch_unwind(AssertUnwindSafe(|| test_function()));
-
-        res.unwrap_or_else(|_| example_result::FAILED_RES)
+impl<'a, T> Runner<'a, T>
+    where T: 'a + Clone + ::std::fmt::Debug
+{
+    pub fn run(mut self) -> ContextReport {
+        let suite = mem::replace(&mut self.suite, None).expect("Expected context");
+        panic::set_hook(Box::new(|_panic_info| {
+            // silently swallows panics
```

Comment:
Oh ! Okay, so then maybe change the comment to: `// XXX panics already catched at the test call site, we want to discard the trace` ?